### ISSUE TITLE
Quick Review

### ISSUE THREAD

**Issue by mratsim** [elijahr/lockfreequeues#1]
Hello there, thanks for your interest in Nim!

A quick review of your code:

## Padding

Instead of padding with arrays

https://github.com/elijahr/lockfreequeues/blob/9f5bc3059cb1f9e40385df14c5d5b2a54f557cbd/src/lockfreequeues/spsc/queueinterface.nim#L25-L30

you can do

```Nim
type
  SPSCQueueInterface* = object
    head* {.align: CACHELINE_BYTES.}: Atomic[uint]
    tail* {.align: CACHELINE_BYTES.}: Atomic[uint]
```

I recommend that you pad the start of the data structure as well since queues are often used in an arrays of queues:

https://github.com/elijahr/lockfreequeues/blob/9f5bc3059cb1f9e40385df14c5d5b2a54f557cbd/src/lockfreequeues/spsc/sharedqueue.nim#L20-L26

## Sizing

Currently your buffer requires being a power of 2, this wastes memory for large buffers
Instead, to distinguish between the empty and the full queue, you can have your capacity run from 0 ..< 2*RealCapacity

See https://github.com/mratsim/weave/blob/master/weave/cross_thread_com/channels_spsc.md#litterature
- http://www.vitorian.com/x1/archives/370
- p3: http://www.sunburst-design.com/papers/CummingsSNUG2002SJ_FIFO1.pdf

And implementation in Nim for a serial queue: https://github.com/mratsim/weave/blob/2e528bd2/weave/datatypes/bounded_queues.nim

## Misc

You don't need to return a seq here, you can return an array, it will be copied

https://github.com/elijahr/lockfreequeues/blob/9f5bc3059cb1f9e40385df14c5d5b2a54f557cbd/src/lockfreequeues/spsc/staticqueue.nim#L89-L102

Though the state might have torn copies.

**Comment by elijahr** [comment]
@mratsim Thank you so much for the welcome and review! I've been enjoying learning Nim over the last few months.

I'll address this feedback and put together a new release sometime in the next few days.

I am not familiar with the term "torn copies" - by that do you mean that `state()` will potentially return invalid state due to the lack of a mutex around head/tail/storage operations? If so - I think that's okay, as the `state()` proc is primarily used to validate the queue state after operations in the single-threaded unit tests. I should add a comment on the proc to clarify that. Or perhaps the proc should instead reside in the unit tests.

Also, I appreciated your talk at NimConf on multithreading. I didn't know about `-fsanitize=thread`, very useful. Cheers!

**Comment by elijahr** [comment]
@mratsim if you'd like to take a look, I think I've addressed your feedback via #3. Cheers!
